### PR TITLE
[GVNSink] Fix incorrect codegen with respect to GEPs #85333

### LINKED ISSUE #85333: GVNSink prodices incorrect codegen with respect to GEPs
```llvm
target datalayout = "e-m:e-p:32:32-Fi8-i64:64-v128:64:128-a:0:32-n32-S64"
target triple = "thumbv7em-none-unknown-eabi"

%"struct.std::pair" = type <{ i32, %struct.substruct, [2 x i8] }>
%struct.substruct = type { i8, i8 }
%"struct.std::random_access_iterator_tag" = type { i8 }

define linkonce_odr dso_local void @__advance(ptr noundef nonnull align 4 dereferenceable(4) %__i, i32 noundef %__n) local_unnamed_addr {
entry:
  %0 = call i1 @llvm.is.constant.i32(i32 %__n)
  %cmp = icmp eq i32 %__n, 1
  %or.cond = and i1 %0, %cmp
  br i1 %or.cond, label %if.then, label %if.else

if.then:                                          ; preds = %entry
  %1 = load ptr, ptr %__i, align 4
  %incdec.ptr = getelementptr inbounds i8, ptr %1, i32 8
  store ptr %incdec.ptr, ptr %__i, align 4
  br label %if.end6

if.else:                                          ; preds = %entry
  %2 = call i1 @llvm.is.constant.i32(i32 %__n)
  %cmp2 = icmp eq i32 %__n, -1
  %or.cond7 = and i1 %2, %cmp2
  br i1 %or.cond7, label %if.then3, label %if.else5

if.then3:                                         ; preds = %if.else
  %3 = load ptr, ptr %__i, align 4
  %incdec.ptr4 = getelementptr inbounds i8, ptr %3, i32 -8
  store ptr %incdec.ptr4, ptr %__i, align 4
  br label %if.end6

if.else5:                                         ; preds = %if.else
  %4 = load ptr, ptr %__i, align 4
  %add.ptr = getelementptr inbounds %"struct.std::pair", ptr %4, i32 %__n
  store ptr %add.ptr, ptr %__i, align 4
  br label %if.end6

if.end6:                                          ; preds = %if.then3, %if.else5, %if.then
  ret void
}
```

The above IR when piped through gvn-sink via `./bin/opt < /tmp/repro.ll -passes=gvn-sink -S` produces:

```llvm
; ModuleID = '<stdin>'
source_filename = "<stdin>"
target datalayout = "e-m:e-p:32:32-Fi8-i64:64-v128:64:128-a:0:32-n32-S64"
target triple = "thumbv7em-none-unknown-eabi"

define linkonce_odr dso_local void @__advance(ptr noundef nonnull align 4 dereferenceable(4) %__i, i32 noundef %__n) local_unnamed_addr {
entry:
  %0 = call i1 @llvm.is.constant.i32(i32 %__n)
  %cmp = icmp eq i32 %__n, 1
  %or.cond = and i1 %0, %cmp
  br i1 %or.cond, label %if.then, label %if.else

if.then:                                          ; preds = %entry
  br label %if.end6

if.else:                                          ; preds = %entry
  %1 = call i1 @llvm.is.constant.i32(i32 %__n)
  %cmp2 = icmp eq i32 %__n, -1
  %or.cond7 = and i1 %1, %cmp2
  br i1 %or.cond7, label %if.then3, label %if.else5

if.then3:                                         ; preds = %if.else
  br label %if.end6

if.else5:                                         ; preds = %if.else
  br label %if.end6

if.end6:                                          ; preds = %if.else5, %if.then3, %if.then
  %.sink1 = phi i32 [ 8, %if.then ], [ -8, %if.then3 ], [ %__n, %if.else5 ]
  %2 = load ptr, ptr %__i, align 4
  %incdec.ptr = getelementptr inbounds i8, ptr %2, i32 %.sink1
  store ptr %incdec.ptr, ptr %__i, align 4
  ret void
}

; Function Attrs: convergent nocallback nofree nosync nounwind willreturn memory(none)
declare i1 @llvm.is.constant.i32(i32) #0

attributes #0 = { convergent nocallback nofree nosync nounwind willreturn memory(none) }
```

The GEP in the result is incorrect. Prior, the GEPs were
```llvm
  %incdec.ptr = getelementptr inbounds i8, ptr %1, i32 8
  %incdec.ptr4 = getelementptr inbounds i8, ptr %3, i32 -8
  %add.ptr = getelementptr inbounds %"struct.std::pair", ptr %4, i32 %__n
```
but now it's
```llvm
%incdec.ptr = getelementptr inbounds i8, ptr %2, i32 %.sink1
```
This results in an incorrect offset for `%add.ptr = getelementptr inbounds %"struct.std::pair", ptr %4, i32 %__n` since we'd now just add `%.sink1` bytes (where `%.sink1 = %__n` which is the second argument), whereas before we'd add `%__n * 8` bytes (where 8 is the sizeof `%"struct.std::pair"`).

The before ASM is:
```asm
__advance:
        ldr     r2, [r0]
        add.w   r1, r2, r1, lsl #3
        str     r1, [r0]
        bx      lr
```
but the after asm is:
```asm
__advance:
        ldr     r2, [r0]
        add     r1, r2
        str     r1, [r0]
        bx      lr
```

The GEP should not be a candidate for sinking.

### PR DESCRIPTION
**Pull request by hiraditya**
As mentioned in #68882 and https://discourse.llvm.org/t/rfc-replacing-getelementptr-with-ptradd/68699

Gep arithmetic isn't consistent with different types. GVNSink didn't realize this and sank all geps
as long as their operands can be wired via PHIs
in a post-dominator.

Fixes: #85333

### CODE REVIEW

**Review by ilovepi — APPROVED** [review]
Overall, this seems fine to me, but I do wonder about the situation arising in the first place. I was under the impression that GEPs had been canonicalized to always use i8 now when the IR is read in (e.g. auto upgrade). Is that logic in one of the simplification passes? Does that imply that such a pass should always run *before* **GVNSink**? IIRC, what I'm referring to is related to the [ptradd proposal](https://discourse.llvm.org/t/rfc-replacing-getelementptr-with-ptradd/68699).

**Comment by nikic** [comment]
This is probably not the right way to fix this: GEPs are not the only instruction type that has non-operand state. Either there is some other place already taking this into account in GVNSink that needs to be adapted for GEPs, or GVNSink should be calling hasSameSpecialState() (or one of that family of functions).

**Comment by nikic** [comment]
> Overall, this seems fine to me, but I do wonder about the situation arising in the first place. I was under the impression that GEPs had been canonicalized to always use i8 now when the IR is read in (e.g. auto upgrade). Is that logic in one of the simplification passes? Does that imply that such a pass should always run _before_ **GVNSink**? IIRC, what I'm referring to is related to the [ptradd proposal](https://discourse.llvm.org/t/rfc-replacing-getelementptr-with-ptradd/68699).

GVNSink still has to be correct when considered as a standalone pass. Besides, the i8 canonicalization is currently only applied to constant offsets, while GVNSink presumably also operates on non-constants.

File: llvm/test/Transforms/GVNSink/different-gep-types.ll
Comment (nikic) [inline llvm/test/Transforms/GVNSink/different-gep-types.ll:1]: As a warning: GVNSink is non-deterministic, which makes it easy for tests written in one environment to fail on some buildbots. I don't know whether it will affect this specific test. This non-determinism (and the fact that it prevents us from writing tests) has been blocking various fixing in this incredibly broken pass.

File: llvm/test/Transforms/GVNSink/different-gep-types.ll [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+%struct.substruct = type { i8, i8 }
+%"struct.std::random_access_iterator_tag" = type { i8 }
+
+; CHECK: if.end6
+; CHECK: %incdec.ptr.sink = phi ptr [ %incdec.ptr, %if.then ], [ %incdec.ptr4, %if.then3 ], [ %add.ptr, %if.else5 ]
+; CHECK-NEXT: store ptr %incdec.ptr.sink, ptr %__i, align 4
+
+define linkonce_odr dso_local void @__advance(ptr noundef nonnull align 4 dereferenceable(4) %__i, i32 noundef %__n) local_unnamed_addr {
+entry:
+  %0 = call i1 @llvm.is.constant.i32(i32 %__n)
```
Comment (nikic) [inline llvm/test/Transforms/GVNSink/different-gep-types.ll:16]: This test looks very non-minimal.

**Comment by ilovepi** [comment]
> GVNSink still has to be correct when considered as a standalone pass. 

Agreed. That was kind of what I was getting at, though I didn't express it that clearly. I guess my sentiment is "hey, if GVNSink needs the IR to be in a certain state, shouldn't it ensure it's in that state itself (e.g., by running some canonicalization routine)?". 

> Besides, the i8 canonicalization is currently only applied to constant offsets, while GVNSink presumably also operates on non-constants.

Ah, that is true, I had forgotten we *only* do that for constant offsets.

File: llvm/test/Transforms/GVNSink/different-gep-types.ll
Comment (ilovepi) [reply llvm/test/Transforms/GVNSink/different-gep-types.ll:1]: @nikic I filed https://github.com/llvm/llvm-project/issues/77852 a while back when we ran into this, but are there other known issues of non-determinism that you can point me to? This sounds like something worth fixing, since we've found GVNHoist/GVNSink save a decent amount of size for embedded targets, and it would be ideal if we could stabilize it enough to enable in the `-Oz` pipeline by default.

File: llvm/test/Transforms/GVNSink/different-gep-types.ll (hunk 1 above)
Comment (hiraditya) [reply llvm/test/Transforms/GVNSink/different-gep-types.ll]: Updated.

**Comment by hiraditya** [comment]
> This is probably not the right way to fix this: GEPs are not the only instruction type that has non-operand state. Either there is some other place already taking this into account in GVNSink that needs to be adapted for GEPs, or GVNSink should be calling hasSameSpecialState() (or one of that family of functions).

i see. this makes it simpler. Thanks!

File: llvm/lib/Transforms/Scalar/GVNSink.cpp [hunk 2]
```diff
@@ -719,21 +719,22 @@ GVNSink::analyzeInstructionForSinking(LockstepReverseIterator &LRI,
...
     return std::nullopt;
 
   for (unsigned OpNum = 0, E = I0->getNumOperands(); OpNum != E; ++OpNum) {
     ModelledPHI PHI(NewInsts, OpNum, ActivePreds);
     if (PHI.areAllIncomingValuesSame())
       continue;
-    if (!canReplaceOperandWithVariable(I0, OpNum))
-      // We can 't create a PHI from this instruction!
-      return std::nullopt;
+    for (auto &Candidate : NewInsts) {
```
Comment (nikic) [inline llvm/lib/Transforms/Scalar/GVNSink.cpp:733]: This change seems unrelated -- does it have test coverage?

File: llvm/test/Transforms/GVNSink/different-gep-types.ll [hunk 3]
```diff
@@ -0,0 +1,102 @@
+; NOTE: Assertions have been autogenerated by utils/update_test_checks.py UTC_ARGS: --version 4
+; RUN: opt -passes=gvn-sink -S %s | FileCheck %s
+
+target datalayout = "e-m:e-p:32:32-Fi8-i64:64-v128:64:128-a:0:32-n32-S64"
+target triple = "thumbv7em-none-unknown-eabi"
```
Comment (nikic) [inline llvm/test/Transforms/GVNSink/different-gep-types.ll:5]: Can you drop the triple? If not, then you'll have to add REQUIRES.

File: llvm/lib/Transforms/Scalar/GVNSink.cpp (hunk 2 above)
Comment (hiraditya) [reply llvm/lib/Transforms/Scalar/GVNSink.cpp]: Upon further inspection, it doesn't appear to be needed. After checking all instructions represent same operation, checking w.r.t one instruction is sufficient.

File: llvm/lib/Transforms/Scalar/GVNSink.cpp [hunk 4]
```diff
@@ -719,12 +719,11 @@ GVNSink::analyzeInstructionForSinking(LockstepReverseIterator &LRI,
   // try and continue making progress.
   Instruction *I0 = NewInsts[0];
 
-  // If all instructions that are going to participate don't have the same
-  // number of operands, we can't do any useful PHI analysis for all operands.
-  auto hasDifferentNumOperands = [&I0](Instruction *I) {
-    return I->getNumOperands() != I0->getNumOperands();
+  auto hasDifferentOperands = [&I0](Instruction *I) {
```
Comment (nikic) [inline llvm/lib/Transforms/Scalar/GVNSink.cpp:722]: nit: I'd rename this to something like...
```suggestion
  auto isNotSameOperation = [&I0](Instruction *I) {
```
...as the (non-type) operands may still be different.

**Review by nikic — APPROVED** [review]
LGTM

**Comment by dyung** [comment]
@hiraditya, the test you added seems to be failing on windows bots:

- https://lab.llvm.org/buildbot/#/builders/216/builds/38357
- https://lab.llvm.org/buildbot/#/builders/233/builds/9396
- https://lab.llvm.org/buildbot/#/builders/235/builds/7335

Can you take a look and revert if you need time to investigate?

**Comment by hiraditya** [comment]
let me revert it first.

**Comment by nikic** [comment]
This is exactly the test non-determinism issue I've mentioned before. I think at this point we cannot make any fixes to GVNSink until the non-determinism issue is fixed and we can actually write tests for this pass.

**Comment by hiraditya** [comment]
makes sense. i'll update my investigation in the issue (#77852) created by @ilovepi
